src/main/scala/chisel3/util/MemInterface.scala

```diff
@@ -0,0 +1,254 @@
+package chisel3.util
+
+import chisel3._
+
+import chisel3.internal.Builder
+import chisel3.experimental.SourceInfo
+import chisel3.internal.sourceinfo.{MemTransform, SourceInfoTransform}
+import scala.language.reflectiveCalls
+import scala.language.experimental.macros
+
+/** A bundle of signals representing a memory read port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  */
+class MemoryReadPort[T <: Data](tpe: T, addrWidth: Int) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val data = Output(tpe)
+}
+
+/** A bundle of signals representing a memory write port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param masked Whether this read/write port should have an optional mask.
+  *
+  * @note `masked` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `masked`.
+  */
+class MemoryWritePort[T <: Data] private[chisel3] (tpe: T, addrWidth: Int, masked: Boolean) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val data = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (masked) {
+    val maskSize = tpe match {
+      case vec: Vec[_] => vec.size
+      case _ => 0
+    }
+    Some(Input(Vec(maskSize, Bool())))
+  } else {
+    None
+  }
+}
+
+/** A bundle of signals representing a memory read/write port.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param addrWidth The width of the address signal
+  * @param masked Whether this read/write port should have an optional mask.
+  *
+  * @note `masked` is only valid if tpe is a Vec; if this is not the case no mask will be initialized
+  *       regardless of the value of `masked`.
+  */
+class MemoryReadWritePort[T <: Data] private[chisel3] (tpe: T, addrWidth: Int, masked: Boolean) extends Bundle {
+  val addr = Input(UInt(addrWidth.W))
+  val enable = Input(Bool())
+  val isWrite = Input(Bool())
+  val readData = Output(tpe)
+  val writeData = Input(tpe)
+  val mask: Option[Vec[Bool]] = if (masked) {
+    val maskSize = tpe match {
+      case vec: Vec[_] => vec.size
+      case _ => 0
+    }
+    Some(Input(Vec(maskSize, Bool())))
+  } else {
+    None
+  }
+}
+
+/** A IO bundle of signals connecting to the ports of a wrapped `SyncReadMem`, as requested by
+  * `MemInterface.apply`.
+  *
+  * @tparam tpe The data type of the memory port
+  * @param width The width of the address wires of each port
+  * @param numRd The number of read ports
+  * @param numWr The number of write ports
+  * @param numRdWr The number of read/write ports
+  */
+class MemInterface[T <: Data](tpe: T, addrWidth: Int, numRd: Int, numWr: Int, numRdWr: Int, masked: Boolean)
```

Comment:
I'm still wondering, where are clocks.
Basically for each port they should have their own `clock`, otherwise, how do you implement a Async Queue with dual port memory?